<a href="https://colab.research.google.com/github/tjwei/DataScienceSoftware2026/blob/main/pandas01.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [1]:
import pandas as pd

In [2]:
pd.read_html('https://en.wikipedia.org/wiki/List_of_countries_and_dependencies_by_population')

HTTPError: HTTP Error 403: Forbidden

In [3]:
!curl -JLO https://en.wikipedia.org/wiki/List_of_countries_and_dependencies_by_population

  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100 1352k  100 1352k    0     0  13.5M      0 --:--:-- --:--:-- --:--:-- 13.6M


In [4]:
tables = pd.read_html('List_of_countries_and_dependencies_by_population')
len(tables)

3

### 解決 403 錯誤：加入 User-Agent
透過 `storage_options` 參數，我們可以自訂 HTTP 標頭（Headers），讓網站認為這是一般瀏覽器發出的請求。

In [5]:
import pandas as pd

url = 'https://en.wikipedia.org/wiki/List_of_countries_and_dependencies_by_population'

headers = {'User-Agent': 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/91.0.4472.124 Safari/537.36'}
tables = pd.read_html(url, storage_options=headers)

print(len(tables))

df = tables[0]
display(df.head())

3


,Location,Population,% of world,Date,Source (official or from the United Nations),Notes
0,World,8.232000e+09,100%,13 Jun 2025,UN projection[1][3],NaN
1,India,1.429404e+09,17.3%,1 Jul 2026,Official projection[4],[b]
2,China,1.404890e+09,17.0%,31 Dec 2025,Official estimate[5],[c]
3,United States,3.417849e+08,4.1%,1 Jul 2025,Official estimate[6],[d]
4,Indonesia,2.883151e+08,3.5%,31 Dec 2025,National annual projection[7],NaN


In [6]:
df.shape

(241, 6)

In [7]:
df.dtypes

,0
Location,object
Population,float64
% of world,object
Date,object
Source (official or from the United Nations),object
Notes,object


In [8]:
df

,Location,Population,% of world,Date,Source (official or from the United Nations),Notes
0,World,8.232000e+09,100%,13 Jun 2025,UN projection[1][3],NaN
1,India,1.429404e+09,17.3%,1 Jul 2026,Official projection[4],[b]
2,China,1.404890e+09,17.0%,31 Dec 2025,Official estimate[5],[c]
3,United States,3.417849e+08,4.1%,1 Jul 2025,Official estimate[6],[d]
4,Indonesia,2.883151e+08,3.5%,31 Dec 2025,National annual projection[7],NaN
...,...,...,...,...,...,...
236,Christmas Island (Australia),1.692000e+03,0%,1 Jan 2021,2021 Census[247],NaN
237,Niue (New Zealand),1.681000e+03,0%,11 Nov 2022,2022 Census[248],NaN
238,Vatican City,8.820000e+02,0%,31 Dec 2024,Official figure[249],[ah]
239,Cocos (Keeling) Islands (Australia),5.930000e+02,0%,30 Jun 2020,2021 Census[250],NaN


In [9]:
df = df.dropna(how='all')

In [10]:
df.Population.isna().any()

np.False_

In [11]:
df.shape

(240, 6)

In [12]:
df.Location

,Location
0,World
1,India
2,China
3,United States
4,Indonesia
...,...
236,Christmas Island (Australia)
237,Niue (New Zealand)
238,Vatican City
239,Cocos (Keeling) Islands (Australia)


In [14]:
pd.read_

'2.2.3'

In [15]:
!pip list|grep pandas

geopandas                             1.2.0
pandas                                2.2.3
pandas-datareader                     0.11.1
pandas-gbq                            0.30.0
pandas-stubs                          2.2.3.250527
sklearn-pandas                        2.2.0


In [16]:
df[df['Location'].str.startswith('Ta')]

,Location,Population,% of world,Date,Source (official or from the United Nations),Notes
21,Tanzania,70038850.0,0.8%,1 Jul 2026,Official projection[25],[h]
60,Taiwan,23224721.0,0.3%,31 Aug 2026,Official estimate[72],[p]
89,Tajikistan,10721000.0,0.1%,1 Jan 2026,Official estimate[100],NaN


In [17]:
import humanize
humanize.number.intcomma(
    df.loc[1:].Population.mean()
    )

'33,780,573.43096235'

In [ ]:
df[1:].Population.quantile([0.25,0.5,0.75, 0.8])

In [ ]:
from matplotlib import pyplot as plt

In [ ]:
plt.hist(df[1:].Population, bins=100);

In [ ]:
df[df.Population > 100_000_000]

In [ ]:
df[df.Population > 1000_000_000].Population[1:].sum() / df.Population[0]

In [ ]:
type(df.Population)

In [ ]:
df['Location'].tail(20)

In [ ]:
df.loc[0]

In [ ]:
df.loc[0].index

In [ ]:
df['Population']

In [ ]:
df.T

In [ ]:
(df.T)[0]

In [ ]:
df[['Population', 'Population', 'Location', ]]

In [ ]:
df.loc[[100,11,1,236]]

In [20]:
import numpy as np
s = pd.Series(np.array([1, 3, 5, np.nan, 6, 8]))
s

,0
0,1.0
1,3.0
2,5.0
3,NaN
4,6.0
5,8.0


In [21]:
pd.Series(np.random.randn(5), index=['a', 'b', 'c', 'd', 'e'])

,0
a,1.284474
b,-0.899755
c,-0.115051
d,0.256425
e,-0.830508


In [23]:
import numpy as np
import pandas as pd

degrees = [30, 45, 90]
sine_values = np.sin(np.deg2rad(degrees))

sine_table = pd.Series(sine_values, index=degrees)
sine_table

,0
30,0.500000
45,0.707107
90,1.000000


In [24]:
import numpy as np
import pandas as pd

# Adding more common angles
degrees = [0, 30, 45, 60, 90, 180, 270, 360]
sine_values = np.sin(np.deg2rad(degrees))

# Clean up floating point precision issues (e.g. sin(180) should be 0)
sine_values = np.round(sine_values, 10)

sine_table_expanded = pd.Series(sine_values, index=degrees)
sine_table_expanded

,0
0,0.000000
30,0.500000
45,0.707107
60,0.866025
90,1.000000
180,0.000000
270,-1.000000
360,-0.000000


In [51]:
dates = pd.date_range("20261007", periods=6)
dates.values

array(['2026-10-07T00:00:00.000000000', '2026-10-08T00:00:00.000000000',
       '2026-10-09T00:00:00.000000000', '2026-10-10T00:00:00.000000000',
       '2026-10-11T00:00:00.000000000', '2026-10-12T00:00:00.000000000'],
      dtype='datetime64[ns]')

In [71]:
df = pd.DataFrame(np.random.randn(6, 4), index=dates.values, columns=['a','b','c','d'])

In [72]:
df

,a,b,c,d
2026-10-07,-0.533932,0.476467,1.266895,-1.565196
2026-10-08,0.345166,1.509139,0.487505,0.224264
2026-10-09,-0.748994,2.043132,-1.801740,0.895932
2026-10-10,0.426549,0.651836,-1.177126,1.745680
2026-10-11,0.057504,0.234897,-1.036472,-2.126752
2026-10-12,-1.239725,-1.111705,-0.102706,-0.691303


In [56]:
df2 = pd.DataFrame(
    {
        "A": 1.0,
        "B": pd.Timestamp("20130102"),
        "C": pd.Series(1, index=list(range(4)), dtype="float32"),
        "D": np.array([3] * 4, dtype="int32"),
        "E": pd.Categorical(["test", "train", "test", "train"]),
        "F": "foo",
    }
)


In [62]:
df2.T.dtypes

,0
0,object
1,object
2,object
3,object


In [63]:
import pandas as pd

# Create a simple dictionary
data = {
    'Name': ['Alice', 'Bob', 'Charlie'],
    'Age': [25, 30, 35],
    'City': ['New York', 'London', 'Paris']
}

# Construct DataFrame using from_dict
df_dict = pd.DataFrame.from_dict(data)
df_dict

,Name,Age,City
0,Alice,25,New York
1,Bob,30,London
2,Charlie,35,Paris


In [67]:
import pandas as pd

# Create a dictionary where we want the keys to be rows
data = {
    'Row_A': [1, 'a', 3],
    'Row_B': [4, 'b', 6],
    'Row_C': [7, 'c', 9]
}

# Construct DataFrame using from_dict with orient='index'
df_orient = pd.DataFrame.from_dict(data, orient='index', columns=['Col_1', 'Col_2', 'Col_3'])
df_orient

,Col_1,Col_2,Col_3
Row_A,1,a,3
Row_B,4,b,6
Row_C,7,c,9


In [64]:
import pandas as pd

# Create a list of tuples representing records
records = [
    (1, 'Alice', 'Engineer'),
    (2, 'Bob', 'Designer'),
    (3, 'Charlie', 'Artist')
]

# Construct DataFrame using from_records and specify column names
df_records = pd.DataFrame.from_records(records, columns=['ID', 'Name', 'Role'])
df_records

,ID,Name,Role
0,1,Alice,Engineer
1,2,Bob,Designer
2,3,Charlie,Artist


In [69]:
df.to_numpy()

array([[-0.57132353, -0.52996723,  1.63520444, -0.70653869],
       [-0.16219599, -0.1228071 ,  0.52642194, -0.23573774],
       [ 1.85999804, -0.37887551,  0.42187951, -0.81482853],
       [ 1.50839033, -0.38815679,  0.40911425,  0.18984517],
       [-0.90570682,  0.09716007,  0.3103628 , -1.13881718],
       [ 1.40682475, -0.81439553,  0.54262341, -0.13329623]])

In [70]:
df.values

array([[-0.57132353, -0.52996723,  1.63520444, -0.70653869],
       [-0.16219599, -0.1228071 ,  0.52642194, -0.23573774],
       [ 1.85999804, -0.37887551,  0.42187951, -0.81482853],
       [ 1.50839033, -0.38815679,  0.40911425,  0.18984517],
       [-0.90570682,  0.09716007,  0.3103628 , -1.13881718],
       [ 1.40682475, -0.81439553,  0.54262341, -0.13329623]])

In [99]:
df = pd.DataFrame(np.random.randn(6, 4), index=dates.values, columns=['a','b','c','d'])
df

,a,b,c,d
2026-10-07,-0.491993,0.372788,0.287851,1.302490
2026-10-08,1.149600,-0.962406,1.258133,-1.162349
2026-10-09,-0.028482,0.219304,1.175020,0.346160
2026-10-10,-1.568949,-0.438998,0.015266,0.386569
2026-10-11,0.256246,1.533076,1.520994,-1.393985
2026-10-12,1.260359,0.011981,-1.191553,-0.824658


In [76]:
df.loc['2026-10-8':'2026-10-11']

,a,b,c,d
2026-10-08,-0.007596,-1.101149,-0.688045,-0.249796
2026-10-09,-0.965306,-0.484330,-0.525757,-0.379421
2026-10-10,0.096305,-1.136206,0.804640,-0.555662
2026-10-11,0.140729,1.326791,1.112832,-0.494450


In [86]:
df2 = df.loc[['2026-10-12','2026-10-9','2026-10-07','2026-10-8','2026-10-11']].copy()
df2

,a,b,c,d
2026-10-12,-0.940473,0.643538,-0.503280,0.009432
2026-10-09,-0.965306,-0.484330,-0.525757,-0.379421
2026-10-07,-0.873419,-0.238330,-0.350945,1.854248
2026-10-08,-0.007596,-1.101149,-0.688045,-0.249796
2026-10-11,0.140729,1.326791,1.112832,-0.494450


In [90]:
df2.loc['2026-10-09':]

,a,b,c,d
2026-10-12,-0.940473,0.643538,-0.503280,0.009432
2026-10-09,-0.965306,-0.484330,-0.525757,-0.379421
2026-10-11,0.140729,1.326791,1.112832,-0.494450


In [101]:
df_copy = df.copy()

In [110]:
df = df_copy.copy()
df['F'] = pd.Series([1,2])#, index=[dates[3], dates[1]])
df


,a,b,c,d,F
2026-10-07,-0.491993,0.372788,0.287851,1.302490,NaN
2026-10-08,1.149600,-0.962406,1.258133,-1.162349,NaN
2026-10-09,-0.028482,0.219304,1.175020,0.346160,NaN
2026-10-10,-1.568949,-0.438998,0.015266,0.386569,NaN
2026-10-11,0.256246,1.533076,1.520994,-1.393985,NaN
2026-10-12,1.260359,0.011981,-1.191553,-0.824658,NaN


In [106]:
pd.Series([1,2,3,4,5,6])

,0
0,1
1,2
2,3
3,4
4,5
5,6
